### DISCUSSION TITLE
Assign array of BaseVariableType to Event property

### DISCUSSION THREAD

**Discussion by hirschenberger** [FreeOpcUa/opcua-asyncio#1731] [Q&A]
Hi, I have to assign an Array of BaseVariableTypes to an Event's property.

The Variable type and Event types are coming from an external XML definition. They look like this:

![image](https://github.com/user-attachments/assets/2ccd9ea4-f1fc-443d-ac37-2c01aca68651)

![image](https://github.com/user-attachments/assets/3ee44ee1-bd36-4d05-93fa-c79cda97a22a)

When I want to assign the array of objects to the property like this:

```python
ev_gen.event.AxesTemperatures = [ua.Variant(obj1, ua.VariantType.ExtensionObject)]
```

I get the following error:

```
WARNING:asyncua.common.copy_node_util:Instantiate: while copying attributes from node type ns=5;i=2011, attribute AccessLevel, statuscode is StatusCode(value=2150957056)
WARNING:asyncua.common.copy_node_util:Instantiate: while copying attributes from node type ns=5;i=2011, attribute UserAccessLevel, statuscode is StatusCode(value=2150957056)
WARNING:asyncua.common.copy_node_util:Instantiate: while copying attributes from node type ns=5;i=2011, attribute MinimumSamplingInterval, statuscode is StatusCode(value=2150957056)
WARNING:asyncua.common.copy_node_util:Instantiate: while copying attributes from node type ns=5;i=2011, attribute Historizing, statuscode is StatusCode(value=2150957056)
ERROR:ti.db.web:cannot pickle '_queue.SimpleQueue' object
```

I'd be happy for any hints and tips.

**Comment by schroeder-** [comment]
```
# init code
myobj = ua.Node('....') # Emitting node
etype = ua.Node('....') # NodeId of the AxesTemperatures event 
myevgen = await server.get_event_generator(etype, myobj) 

# generate event
myevgen.event.Message = ua.LocalizedText("MyFirstEvent %d" % count)
myevgen.event.Severity = ua.UInt(100)
myevgen.event.AxisNumber = ua.Int(1)
...
await myevgen.trigger()
```

**Reply by hirschenberger** [reply]
The Event type is not AxisTemperatures, it's TrainAxisTemperatureEventType with a 1-Dim Property AxisTemperatures that should contain the AxisTemperatureType objects.

**Reply by schroeder-** [reply]
My bad.
`ev_gen.event.AxesTemperatures = [ua.AxisTemperatureType(AxisNumber=..., TemperatureBoxLeft=...,...), ua.AxisTemperatureType(...)]`

**Reply by hirschenberger** [reply]
Sorry, I don't know how this should work. It says `module 'asyncua.ua' has no attribute 'AxisTemperatureType'`

Should this type be put into the ua namespace automagically after reading the XML file?

**Reply by schroeder-** [reply]
Yes after importing the XML file(`server.import_xml(...)`), the class should be in the `asyncua.ua` module. This is the way to instantiate OPC UA Datatypes.

**Reply by hirschenberger** [reply]
Ok, thank you!

**Reply by hirschenberger** [reply]
I think I know what's wrong.

The `AxisTemperatures` is an array with `ValueRank=1` and a DataType of `UInteger`, that's how I got the XML-Definition.

I can't change it to the actual `AxisTemperaturesType` b.c. when I load the XML I get an error that the datatype is 'not a subtype of a builtin type'. 

Is this a limitation of OpcUa or opcua-asyncio?

**Reply by AndreasHeine** [reply]
not sure if the type is part of the opc ua core spec but if you load xml files its some times necessary to load the typedefinitions:

https://github.com/FreeOpcUa/opcua-asyncio/blob/ac141d753181dcb18bad5dbd606b60ba82935c6c/asyncua/server/server.py#L828

or for older 1.03 structures: 
https://github.com/FreeOpcUa/opcua-asyncio/blob/ac141d753181dcb18bad5dbd606b60ba82935c6c/asyncua/server/server.py#L819

after loading they should be available while runtime under the "ua" module!

**Reply by hirschenberger** [reply]
Thanks for your reply but that's not issue here. It's more that I can't assign a list of `AxisTemperaturesType` objects to `AxisTemperatures` bc. it's an array and it's `DataType` is specified as `UInteger` with Rank=1.

I can't change the type to `AxisTemperaturesType` with Rank=1 bc. opcua-asyncio does not load the XML if I do so. It seems Array's types have to be of built-in types. The question is, is this a OpcUa limitation? How can I create my List instead?

**Reply by schroeder-** [reply]
You could try this:
`xx.AxisTemperatures = [ua.UInt64(0), ua.UInt64(...)]`
But it looks like we don't support UInteger, because UInteger could be(UInt16, UInt32,...).
Maybe you could post a minimal xml that only contains event and all needed datatypes, so I could try to add the feature.

**Reply by schroeder-** [reply]
If i read it correct as of spec UInteger -> Variant, that should contain an subtype of UInteger(UInt16, UInt32, ...).
So you could try this:
`xx.AxisTemperatures = [ua.Variant(ua.UInt64(0)), ....]`
ot this:
`xx.AxisTemperatures = ua.Variant([ua.UInt64(0)), ...])`

**Reply by hirschenberger** [reply]
Sorry, that's not what I need. I need:

```
xx.AxisTemperatures = [ua.AxisTemperaturesType(..), ua.AxisTemperaturesType(..)...]
```

But when I change the Type of `AxisTemperatures` from `UInteger of Rank=1` to `AxisTemperaturesType of Rank=1` the XML does not load as it says a built-in type is required.

**Reply by schroeder-** [reply]
Then you need to post a minimal xml, so the xmlimport can be expanded. Also you could post a complete log file if you load the xml with the error.

**Reply by hirschenberger** [reply]
I think I found out that there's a flaw in the Spec that we got.

The Datatype of AxisTemperatures has to be a different type than UInteger, but has to be defined in `DataType.BaseDataType`, possibly derived from `Structure`. So it can be used instead of UInteger in an array.

We must change the spec in that regard.

Thanks a lot for your help.
